ludwig/data/negative_sampling.py

```diff
@@ -0,0 +1,106 @@
+import logging
+import time
+from typing import Any, List, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy
+
+from ludwig.utils.types import DataFrame
+
+
+def _negative_sample_user(interaction_row: np.array, neg_pos_ratio: int, extra_samples: int) -> Tuple[List[int], int]:
+    """Returns a list of negative item indices for given user-item interactions.
+
+    If there are not enough negative items, takes all of them and adds the difference to the extra_samples
+    otherwise, samples with replacement.
+
+    Params:
+        interaction_row: user-item interaction row
+        neg_pos_ratio: number of negative samples per positive sample
+        extra_samples: number of additional samples to add to the negative sample list
+    Returns:
+        Tuple of list of negative item indices and number of extra samples
+    """
+    # Find all items that are not interacted with by the user
+    neg_items = np.where(interaction_row == 0)[1]
+    available_samples = len(neg_items)
+
+    # Randomly sample negative items
+    npos = interaction_row.shape[1] - len(neg_items)
+    samples_required = npos * neg_pos_ratio + extra_samples
+    should_sample = samples_required <= available_samples
+
+    neg_items = np.random.choice(neg_items, samples_required, replace=False) if should_sample else neg_items
+
+    return neg_items.tolist(), max(0, samples_required - available_samples)
+
+
+def negative_sample(
+    df: DataFrame,
+    user_id_col: str = "customer_id",
+    item_id_col: str = "article_id",
+    label_col: str = "label",
+    neg_pos_ratio: int = 1,
+    neg_val: Any = 0,
+    log_pct: int = 0,
+):
+    """Negative sampling for implicit feedback datasets.
+
+    Params:
+        df: DataFrame containing user-item interactions
+        user_id_col: column name for user ids
```

Comment:
Is there a difference between user and item s.t. this is most efficient when `|users| < |items|`?  If so, add to doc string.